src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java

```diff
@@ -18,35 +18,52 @@
 package org.mqttbee.api.mqtt.datatypes;
 
 import org.mqttbee.annotations.NotNull;
+import org.mqttbee.annotations.Nullable;
+
+import java.util.function.Function;
 
 /**
  * @author Silvio Giebl
  */
-public class MqttSharedTopicFilterBuilder extends MqttTopicFilterBuilder {
+public class MqttSharedTopicFilterBuilder<P> extends MqttTopicFilterBuilder<P> {
+
+    public static <P> MqttSharedTopicFilterBuilder<P> create(
+            @NotNull final String shareName, @NotNull final String base,
+            @Nullable final Function<? super MqttSharedTopicFilter, P> parentConsumer) {
+
+        return new MqttSharedTopicFilterBuilder<>(shareName, base, (parentConsumer == null) ? null :
+                topicFilter -> parentConsumer.apply((MqttSharedTopicFilter) topicFilter));
+    }
 
     private final String shareName;
 
-    MqttSharedTopicFilterBuilder(@NotNull final String shareName, @NotNull final String base) {
-        super(base);
+    public MqttSharedTopicFilterBuilder(
+            @NotNull final String shareName, @NotNull final String base,
+            @Nullable final Function<? super MqttTopicFilter, P> parentConsumer) {
+
+        super(base, parentConsumer);
         this.shareName = shareName;
     }
 
     @NotNull
     @Override
-    public MqttSharedTopicFilterBuilder subTopic(@NotNull final String subTopic) {
-        return (MqttSharedTopicFilterBuilder) super.subTopic(subTopic);
+    public MqttSharedTopicFilterBuilder<P> addLevel(@NotNull final String subTopic) {
+        super.addLevel(subTopic);
+        return this;
     }
 
     @NotNull
     @Override
-    public MqttSharedTopicFilterBuilder singleLevelWildcard() {
-        return (MqttSharedTopicFilterBuilder) super.singleLevelWildcard();
+    public MqttSharedTopicFilterBuilder<P> singleLevelWildcard() {
+        super.singleLevelWildcard();
+        return this;
     }
 
     @NotNull
     @Override
-    public MqttSharedTopicFilter multiLevelWildcard() {
-        return (MqttSharedTopicFilter) super.multiLevelWildcard();
+    public MqttSharedTopicFilter multiLevelWildcardAndBuild() {
+        super.multiLevelWildcardAndBuild();
+        return build();
```

Comment:
build() is called two times. I think you have to call super.multiLevelWildcard() and make it protected.

Reply:
you are right, casting is maybe more clean
`return (MqttSharedTopicFilter) super.multiLevelWildcardAndBuild()`